workflow/rules/qc.smk

```diff
@@ -0,0 +1,130 @@
+    params:
+        prefix=config['output'] + "{Organism}/{refGenome}/" + config['qcDir'] + "{Organism}_{refGenome}"
+    shell:
+        """
+        vcftools --gzvcf {input.vcf} --FILTER-summary --out {params.prefix}
+        vcftools --gzvcf {input.vcf} --out {params.prefix} --depth
+        vcftools --gzvcf {input.vcf} --out {params.prefix} --missing-indv
+        """
+
+rule subsample_snps:
+    input:
+        vcf = config['output'] + "{Organism}/{refGenome}/" + "{Organism}_{refGenome}.final.vcf.gz"
+    output:
+        filt_snps = config['output'] + "{Organism}/{refGenome}/" + config['qcDir'] + "{Organism}_{refGenome}_filt_snps.txt",
+        tmp_rand_snps = config['output'] + "{Organism}/{refGenome}/" + config['qcDir'] + "{Organism}_{refGenome}_tmp_rand_snps.txt",
+        filtered = config['output'] + "{Organism}/{refGenome}/" + config['qcDir'] + "{Organism}_{refGenome}.filtered.vcf.gz",
+        filtered_idx = config['output'] + "{Organism}/{refGenome}/" + config['qcDir'] + "{Organism}_{refGenome}.filtered.vcf.gz.csi",
+        pruned = config['output'] + "{Organism}/{refGenome}/" + config['qcDir'] + "{Organism}_{refGenome}.pruned.vcf.gz"
+    conda:
+        "../envs/qc.yml"
+    resources:
+        mem_mb = lambda wildcards, attempt: attempt * res_config['vcftools']['mem']    # this is the overall memory requested
+    shell:
+        """
+        bcftools view -v snps -m2 -M2 -f .,PASS -e 'AF==1 | AF==0 | ALT="*" | TYPE~"indel" | ref="N"' {input.vcf} -O z -o {output.filtered}
+        bcftools index {output.filtered}
+        bcftools query -f '%CHROM\t%POS\n' {output.filtered} > {output.filt_snps}
+        sort -R {output.filt_snps}| head -50000 > {output.tmp_rand_snps}
+        bcftools view -R {output.tmp_rand_snps} {output.filtered} -O z -o {output.pruned}
+        """
+
+rule plink:
+    """
+    Call plink PCA.
+    """
+    input:
+        vcf = config['output'] + "{Organism}/{refGenome}/" + config['qcDir'] + "{Organism}_{refGenome}.pruned.vcf.gz"
+        
+    params:
+        prefix=config['output'] + "{Organism}/{refGenome}/" + config['qcDir'] + "{Organism}_{refGenome}",
+        threads = res_config['plink']['threads']
+    output: 
+        bed = config['output'] + "{Organism}/{refGenome}/" + config['qcDir'] + "{Organism}_{refGenome}.bed",
+        bim = config['output'] + "{Organism}/{refGenome}/" + config['qcDir'] + "{Organism}_{refGenome}.bim",
+        eigenvec = config['output'] + "{Organism}/{refGenome}/" + config['qcDir'] + "{Organism}_{refGenome}.eigenvec"
+    resources:
+        mem_mb = lambda wildcards, attempt: attempt * res_config['vcftools']['mem']    # this is the overall memory requested
+    conda:
+        "../envs/qc.yml"
+    shell:
+        #plink 2 for king relatedness matrix (robust to structure) and plink 1.9 for distance matrix
+        """
+        plink2 --vcf {input.vcf} --pca 2 --out {params.prefix} --allow-extra-chr --autosome-num 95 --make-bed --make-king square --const-fid --bad-freqs
```

Comment:
Yeah, not sure what is safest here, and open to suggestions. I am a little unclear on the consequence of the --autosome-num flag, I just set the upper limit. AFAIK it just affects the formatting of the chrom/pos file and still processes all the variants. It does run fine on contig assemblies though, but due to plink's internal naming reqs I am sure there are some names that crash it

Reply:
My understanding is that PLINK will crash if you have a contig/scaffold/chromosome with an entirely numeric name that is greater than the value of the autosome-num parameter. So if for example you have 500 scaffolds and rather than being named scaffold1, scaffold2, ... , scaffold500, they are just named 1, 2, ..., 500, PLINK will crash. 

If you have ANY scaffolds/contigs that have non-numeric names, PLINK will crash unless you include the --allow-extra-chr rule. 

I think for almost all genomes the current command line will be okay. But it will fail for e.g. a PacBio HiFi assembly with 423 contigs that are named 1, 2, 3, ... , 423 instead of contig1, etc. 

This will not matter for any NCBI genome, since they all have alphanumeric scaffold/chr names anyway. So it would only be an issue with a user-provided reference. 

On further thought, the safest thing to do is definitely check this at the very beginning, because if you get to this step and plink throws an error it will be a pain to fix. But could just put this as a known issue to solve later.

Reply:
I set up a simple rule to check (from the fai file) if the contigs are numeric only. The qc plots (and plink commands) are not run if this is the case, and an error is printing saying QC plots are not generated due to numeric only contigs